### ISSUE TITLE
Refactor operator() for genetic value calculations

### ISSUE THREAD

**Issue by molpopgen** [molpopgen/fwdpy11#372]
The current signature looks like this:

```cpp
        virtual void
        operator()(const GSLrng_t& rng, std::size_t diploid_index,
                   const DiploidPopulation& pop, DiploidMetadata& metadata) const
        {
            metadata.g = calculate_gvalue(diploid_index, pop);
            metadata.e = noise(rng, metadata, metadata.parents[0],
                               metadata.parents[1], pop);
            metadata.w = genetic_value_to_fitness(metadata);
        }
```

Something like this would be much better:

```cpp
        virtual void
        operator()(const GSLrng_t& rng, const DiploidPopulation & pop, const DiploidGenotype & offspring, DiploidMetadata & offspring_metadata) const
        {
            // update accordingly.
        }
```

These changes are related to #371, which ultimately has to do with how offspring metadata can be affected by parents.  The current API (see above) doesn't allow parental genotypes to play a role.  

The key here is that the offspring metadata aren't in `pop`, which contains parental metadata.  However, the parent indexes are in the offspring metadata, so a function can check parental genotypes, or at least that's the situation we want.

In order for parental genotypes to affect offspring metadata, the simulation code needs to not swap the DiploidGenotypes out before calculating all the fitnesses.  See comments in with_tree_sequences.cc

**Comment by molpopgen** [comment]
Note: we are closing this in milestone 0.7.0, which is kinda kicking the can down the road in favor of doing other things first that are arguably more useful.

**Comment by molpopgen** [comment]
In light of features planned for 0.9.0, I think this one is important again.

**Comment by molpopgen** [comment]
Closed by #557
